lualib/resty/events/broker.lua

```diff
@@ -28,64 +25,125 @@ local decode = codec.decode
 local cjson_encode = cjson.encode
 
 local MAX_UNIQUE_EVENTS = 1024
+local WEAK_KEYS_MT = { __mode = "k", }
 
-local function is_closed(err)
-    return err and
-           (str_sub(err, -6) == "closed" or
-            str_sub(err, -11) == "broken pipe")
+local function terminating(self, worker_connection)
+    return self._clients[worker_connection] == nil or exiting()
 end
 
--- broadcast to all/unique workers
-local function broadcast_events(self, unique, data)
-    local n = 0
-
-    -- if unique, schedule to a random worker
-    local idx = unique and random(1, nkeys(self._clients))
+local function get_event_data(self, event_data)
+    local unique = event_data.spec.unique
+    if unique then
+        local uniques = self._uniques
+        if uniques:get(unique) then
+            if not exiting() then
+                log(DEBUG, "unique event is duplicate: ", unique)
+            end
 
-    for _, q in pairs(self._clients) do
+            return
+        end
 
-        -- skip some and broadcast to one workers
-        if unique then
-            idx = idx - 1
+        uniques:set(unique, 1, self._opts.unique_timeout)
+    end
+    return event_data.data, unique
+end
 
-            if idx > 0 then
-                goto continue
-            end
-        end
+-- broadcast to all/unique workers
+local function broadcast_events(self, event_data)
+    local data, unique = get_event_data(self, event_data)
+    if not data then
+        return
+    end
 
-        local ok, err = q:push(data)
+    local n = 0
 
-        if not ok then
+    -- pairs is "random" enough for unique
+    for _, client_queue in pairs(self._clients) do
```

Comment:
my understanding is that the order of iteration of `pairs` depends on the value of the hash of each key in the table, so while this is in fact "random" or unpredictable when keys are created/rehashed, doesn't it mean the iteration order will stay the same for the `self._clients` table until then? I.e. would this implementation risk to run "unique" events on the same worker repeatedly?

Reply:
@samugi it will mean a bit like that. Order may change or it may not.  It is undefined. We dit just on't almost never use unique events and what this does is that then runs it in a same worker. Not a huge deal. There is a bigger benefit of it actually retrying it on other worker if it faile on previous. So it is more robust, but does not balance the load the same as before, though I have no idea why we would need that.